rosbag2_compression/src/rosbag2_compression/zstd_decompressor.cpp

```diff
@@ -0,0 +1,240 @@
+// limitations under the License.
+
+#include <chrono>
+#include <cstdio>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "rcpputils/filesystem_helper.hpp"
+
+#include "rosbag2_compression/zstd_decompressor.hpp"
+
+#include "rosbag2_storage/filesystem_helper.hpp"
+
+#include "logging.hpp"
+
+namespace
+{
+
+// String constant used to identify ZstdDecompressor.
+constexpr const char kDecompressionIdentifier[] = "zstd";
+
+/**
+ * Open a file using the C API.
+ * This function calls OS-specific implementation of fopen.
+ *
+ * \param uri is the path to the file
+ * \param read_mode is the read mode string accepted by fopen.
+ * \return the FILE pointer or nullptr if the file was not opened.
+ */
+FILE * open_file(const std::string & uri, const std::string & read_mode)
+{
+  FILE * fp{nullptr};
+#ifdef _WIN32
+  fopen_s(&fp, uri.c_str(), read_mode.c_str());
+#else
+  fp = std::fopen(uri.c_str(), read_mode.c_str());
+#endif
+  return fp;
+}
+
+/**
+ * Read a file from the supplied uri into a vector.
+ *
+ * \param uri is the path to the file.
+ * \return the contents of the buffer as a vector.
+ */
+std::vector<uint8_t> get_input_buffer(const std::string & uri)
+{
+  // Get the file size
+  const auto compressed_buffer_length = rosbag2_storage::FilesystemHelper::get_file_size(uri);
+
+  // Read in buffer, handling accordingly
+  const auto file_pointer = open_file(uri.c_str(), "rb");
+  if (file_pointer == nullptr) {
+    std::stringstream errmsg;
+    errmsg << "Failed to open file: \"" << uri << "\" for binary reading!";
+
+    throw std::runtime_error{errmsg.str()};
+  }
+
+  // Initialize compress_buffer with size = compressed_buffer_length.
+  // Uniform initialization cannot be used here since it will choose
+  // the initializer list constructor instead.
+  std::vector<uint8_t> compressed_buffer(compressed_buffer_length);
+
+  const auto nRead = fread(
+    compressed_buffer.data(), sizeof(decltype(compressed_buffer)::value_type),
+    compressed_buffer_length, file_pointer);
+
+  if (nRead != compressed_buffer_length) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Bytes read !(" <<
+      nRead << ") != compressed_buffer_length (" << compressed_buffer_length <<
+      ")!");
+    // An error indicator is set by fread, so the following check will throw.
+  }
+
+  if (ferror(file_pointer)) {
+    fclose(file_pointer);
+
+    std::stringstream errmsg;
+    errmsg << "Unable to read binary data from file: \"" << uri << "\"!";
+
+    throw std::runtime_error{errmsg.str()};
+  }
+
+  fclose(file_pointer);
+  return compressed_buffer;
+}
+
+/**
+ * Writes the output buffer to the specified file path.
+ * \param output_buffer is the data to write.
+ * \param output_buffer_length is the number of bytes to write.
+ * \param uri is the relative file path to the output storage.
+ */
+void write_output_buffer(
+  const uint8_t * output_buffer,
+  const size_t output_buffer_length,
```

Comment:
I think it's still better at this point for `write_output_buffer()` to have a more C++-like interface, eg. pass in a `const vector<uint8_t> &`.

Reply:
Done.
This might revert in the future when we implement streaming, however we don't need it now so passing `vector` by const ref is better.